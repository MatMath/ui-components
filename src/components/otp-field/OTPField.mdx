---
name: OTP Field
menu: Components
---

import { Playground, Props } from 'docz';
import { OTPField } from './OTPField';

# OTP Field

An OTP Field component showing a generated OTP code and its time left using a timer clock.

## Properties

<Props of={OTPField} />

## Usage

### Basic usage

Examples below show usage of the OTP Field

<Playground className={{preview: 'playground-base'}}>
  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={5000}
    validityEndDate={Date.now() + 5000}
  />

<OTPField
  label={'security code'}
  code={'145236'}
  validityPeriod={10000}
  validityEndDate={Date.now() + 10000}
/>

  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={30000}
    validityEndDate={Date.now() + 30000}
  />
</Playground>

### Delay logic

If the initial `validityEndDate` is not equal to `Date.now() + validityPeriod`, then the clock will automatically catch up the delay.

<Playground className={{preview: 'playground-base'}}>
  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={5000}
    validityEndDate={Date.now() + 2000}
  />

<OTPField
  label={'security code'}
  code={'145236'}
  validityPeriod={30000}
  validityEndDate={Date.now() + 15000}
/>

  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={60000}
    validityEndDate={Date.now() + 20000}
  />
</Playground>

### Alert Threshold

Use the `alertThreshold` prop to change the color to an alert color when the the threshold is reached.

<Playground className={{preview: 'playground-base'}}>
  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={5000}
    validityEndDate={Date.now() + 5000}
    alertThreshold={1000}
  />

<OTPField
  label={'security code'}
  code={'145236'}
  validityPeriod={10000}
  validityEndDate={Date.now() + 10000}
  alertThreshold={2000}
/>

  <OTPField
    label={'security code'}
    code={'145236'}
    validityPeriod={30000}
    validityEndDate={Date.now() + 30000}
    alertThreshold={5000}
  />
</Playground>
