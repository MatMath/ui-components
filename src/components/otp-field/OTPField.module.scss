@use '../../colors/colors';
@use '../../typography/typography';

/* remove edge decorators */
::-ms-clear {
  display: none;
}

.root {
  display: flex;
  flex-direction: column;
  align-items: flex-start;

  .label {
    @include typography.body-small;

    color: colors.$grey-00;
    height: 16px;
    text-transform: uppercase;
    font-size: 10px;
    font-weight: 600;
    margin-bottom: 4px;
  }

  .container {
    position: relative;
    display: flex;
    align-items: center;
    margin-bottom: 4px;
    max-width: 100%;
    width: 100%;

    .code {
      @include typography.mono-medium;

      color: colors.$black;
      margin-right: 12px;
    }
  }
}
