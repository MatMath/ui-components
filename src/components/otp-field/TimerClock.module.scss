@use '../../colors/colors';
@use '../../typography/typography';

.base-timer {
  position: relative;
  height: 14px;
  width: 14px;
  line-height: 1;
  font-size: 16px;
  padding: 2px;
}

.base-timer-circle {
  fill: none;
  stroke: none;
}

.base-timer-path-elapsed {
  stroke-width: 2px;
  stroke: colors.$dash-green-04;
}

.base-timer-path-remaining {
  stroke-width: 2px;
  transform: rotate(90deg);
  transform-origin: center;
  transition: stroke 250ms ease-out;
  transition: stroke-dasharray 1s linear;
  stroke: colors.$mid-green-00;

  &.alert-path {
    stroke: colors.$orange-00;
  }
}

.base-timer-svg {
  transform: scaleX(-1);
}
