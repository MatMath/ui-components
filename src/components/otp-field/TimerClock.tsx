import * as React from 'react';

import { getClassNames } from '@utility/cssUtils';
import styles from './TimerClock.module.scss';

const PERIMITER = 38;

export interface TimerClockProps {
  /**
   * Total time the code should be valid in ms
   */
  validityPeriod: number;
  /**
   * Unix time when the code is no longer valid
   */
  validityEndDate: number;
  /**
   * Remaining time after which the circle's color changes to alert in ms
   */
  alertThreshold?: number;
  /**
   * Callback triggered after the timer expires
   * ie when the validityEndDate is reached
   */
  onTimerEnd?: () => void;
}

export const TimerClock = ({
  validityEndDate,
  validityPeriod,
  alertThreshold = 0,
  onTimerEnd
}: TimerClockProps): JSX.Element => {
  const timeRef = React.useRef<NodeJS.Timeout | null>(null);

  // Initialize the timeLeft to the total possible waiting time.
  const [timeLeft, setTimeLeft] = React.useState(validityPeriod);

  React.useEffect(() => {
    if (timeRef.current && timeLeft < 0) {
      timeRef.current && clearTimeout(timeRef.current);
      onTimerEnd && onTimerEnd();
      return;
    }

    // Set the timeLeft to the real time left to catch up the possible delay
    // Important to have a 'catch up' animation
    if (!timeRef.current) {
      setTimeLeft(validityEndDate - Date.now());
    }

    // Update the timeLeft each second (in sync with the 1s animation of the SVG)
    timeRef.current = setTimeout(() => {
      setTimeLeft(timeLeft - 1000);
    }, 1000);

    return () => {
      timeRef.current && clearTimeout(timeRef.current);
    };
  }, [timeLeft, validityEndDate, onTimerEnd]);

  const circleDashArray = ((timeLeft / validityPeriod) * PERIMITER).toFixed(0);
  const isClockStatusAlert = timeLeft < alertThreshold;

  return (
    <div className={styles.baseTimer}>
      <svg
        className={styles.baseTimerSvg}
        viewBox='0 0 14 14'
        xmlns='http://www.w3.org/2000/svg'
      >
        <g className={styles.baseTimerCircle}>
          <circle className={styles.baseTimerPathElapsed} cx='7' cy='7' r='6' />
          <path
            strokeDasharray={`${circleDashArray} ${PERIMITER}`}
            className={getClassNames(styles.baseTimerPathRemaining, {
              [styles.alertPath]: isClockStatusAlert
            })}
            d='
              M 7, 7
              m -6, 0
              a 6,6 0 0,0 12,0
              a 6,6 0 0,0 -12,0
            '
          ></path>
        </g>
      </svg>
    </div>
  );
};
