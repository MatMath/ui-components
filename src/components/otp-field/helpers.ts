export function formatCode(code: string): string {
  return [code.slice(0, 3), ' ', code.slice(3)].join('');
}
