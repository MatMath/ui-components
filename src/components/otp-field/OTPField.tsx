import * as React from 'react';

import { TimerClock, TimerClockProps } from '@components/otp-field/TimerClock';
import { formatCode } from '@components/otp-field/helpers';
import { useOTPCodeShuffler } from '@components/otp-field/otp-code-shuffler/useCodeShuffler';
import styles from './OTPField.module.scss';

export interface OTPFieldProps extends TimerClockProps {
  label: string;
  /**
   * A 6-character otp code string
   * No space character should be within the code
   * If null is provided, the component will shuffle the code
   */
  code: string | null;
}

export const OTPField = ({
  code,
  label,
  onTimerEnd,
  ...clockProps
}: OTPFieldProps): JSX.Element => {
  const [isShuffling, setIsShuffling] = React.useState(false);

  const resetCode = React.useCallback(() => {
    setIsShuffling(true);
    onTimerEnd && onTimerEnd();
  }, [onTimerEnd]);

  React.useEffect(() => {
    setIsShuffling(!code);
  }, [code]);

  const { shuffledCode } = useOTPCodeShuffler({
    shouldShuffle: isShuffling
  });

  const formattedCode = formatCode(isShuffling || !code ? shuffledCode : code);

  return (
    <div className={styles.root}>
      <span className={styles.label}>{label}</span>
      <div className={styles.container}>
        <div className={styles.code}>{formattedCode}</div>
        <TimerClock
          key={clockProps.validityEndDate}
          onTimerEnd={resetCode}
          {...clockProps}
        />
      </div>
    </div>
  );
};
