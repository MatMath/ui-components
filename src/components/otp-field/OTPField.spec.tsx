import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import { OTPField } from '@components/otp-field/OTPField';

Date.now = jest.fn().mockReturnValue(1592829201510);

const DEFAULT_LABEL = 'security code';
const DEFAULT_CODE = '000000';
const CURRENT_DATE = Date.now();
const VALIDITY_PERIOD = 30000; // Time in ms
const VALIDITY_ENDED_TIME = CURRENT_DATE - 1000; // Time in the past

describe('<OTPField>', () => {
  describe('Global render', () => {
    it('should render default', () => {
      const otpField = mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE + VALIDITY_PERIOD}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(otpField).toMatchSnapshot();
    });
  });

  describe('OTP Clock', () => {
    it('should have a full dash array when no time has been elapsed', () => {
      const otpField = mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE + VALIDITY_PERIOD}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(
        otpField
          .find('path')
          .getDOMNode()
          .getAttribute('stroke-dasharray')
      ).toBe('38 38');
    });

    it('should have an empty dash array when there is no more time left', () => {
      const otpField = mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(
        otpField
          .find('path')
          .getDOMNode()
          .getAttribute('stroke-dasharray')
      ).toBe('0 38');
    });
  });

  describe('OTP Field onTimerEnd', () => {
    it('should trigger the onTimerEnd callback when there is not time left', () => {
      const onTimerEnd = jest.fn();
      mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={VALIDITY_ENDED_TIME}
          validityPeriod={VALIDITY_PERIOD}
          onTimerEnd={onTimerEnd}
        />
      );

      expect(onTimerEnd).toBeCalledTimes(1);
    });

    it('should not trigger the onTimerEnd callback when there is still time left', () => {
      const onTimerEnd = jest.fn();
      mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE + 1000}
          validityPeriod={VALIDITY_PERIOD}
          onTimerEnd={onTimerEnd}
        />
      );

      expect(onTimerEnd).toBeCalledTimes(0);
    });
  });

  describe('OTP Field code shuffler', () => {
    it('should trigger a code shuffle when there is no time left', () => {
      const otpField = mount(
        <OTPField
          code={DEFAULT_CODE}
          label={DEFAULT_LABEL}
          validityEndDate={VALIDITY_ENDED_TIME}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(otpField.find('.code').text()).not.toBe(DEFAULT_CODE);
    });

    it('should trigger a code suffle when the code is null', () => {
      const otpField = mount(
        <OTPField
          code={null}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE + VALIDITY_PERIOD}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(otpField.find('.code').text()).toHaveLength(7);
    });

    it('should not trigger a code suffle and show the code when there is time left', () => {
      const code = '123456';
      const formattedCode = '123 456';
      const otpField = mount(
        <OTPField
          code={code}
          label={DEFAULT_LABEL}
          validityEndDate={CURRENT_DATE + VALIDITY_PERIOD}
          validityPeriod={VALIDITY_PERIOD}
        />
      );

      expect(otpField.find('.code').text()).toBe(formattedCode);
    });
  });
});
