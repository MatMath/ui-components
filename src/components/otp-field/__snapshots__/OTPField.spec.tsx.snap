// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OTPField> Global render should render default 1`] = `
<Component
  code="000000"
  label="security code"
  validityEndDate={1592829231510}
  validityPeriod={30000}
>
  <div
    className="root"
  >
    <span
      className="label"
    >
      security code
    </span>
    <div
      className="container"
    >
      <div
        className="code"
      >
        000 000
      </div>
      <Component
        onTimerEnd={[Function]}
        validityEndDate={1592829231510}
        validityPeriod={30000}
      >
        <div>
          <svg
            viewBox="0 0 14 14"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <circle
                cx="7"
                cy="7"
                r="6"
              />
              <path
                className=""
                d="
              M 7, 7
              m -6, 0
              a 6,6 0 0,0 12,0
              a 6,6 0 0,0 -12,0
            "
                strokeDasharray="38 38"
              />
            </g>
          </svg>
        </div>
      </Component>
    </div>
  </div>
</Component>
`;
